game-app/game-core/src/main/java/games/strategy/triplea/delegate/move/validation/MoveValidator.java

```diff
@@ -1450,62 +1452,67 @@ private MoveValidationResult validateParatroops(
         }
       }
-      final Territory routeEnd = route.getEnd();
-      for (final Unit paratroop : paratroopsRequiringTransport) {
+      final boolean friendlyEnd = Matches.isTerritoryFriendly(player).test(route.getEnd());
+      final boolean canMoveNonCombat =
+          Properties.getParatroopersCanMoveDuringNonCombat(data.getProperties());
+      final boolean isWrongPhase = !isNonCombat && friendlyEnd && canMoveNonCombat;
+      final boolean mustAdvanceToBattle = friendlyEnd && !canMoveNonCombat;
+      for (final Unit paratroop : paratroopersToAirTransports.keySet()) {
         if (paratroop.hasMoved()) {
           result.addDisallowedUnit("Cannot paratroop units that have already moved", paratroop);
         }
-        if (Matches.isTerritoryFriendly(player).test(routeEnd)
-            && !Properties.getParatroopersCanMoveDuringNonCombat(data.getProperties())) {
+        if (mustAdvanceToBattle) {
           result.addDisallowedUnit("Paratroops must advance to battle", paratroop);
         }
-        if (!nonCombat
-            && Matches.isTerritoryFriendly(player).test(routeEnd)
-            && Properties.getParatroopersCanMoveDuringNonCombat(data.getProperties())) {
+        if (isWrongPhase) {
           result.addDisallowedUnit(
               "Paratroops may only airlift during Non-Combat Movement Phase", paratroop);
         }
       }
-      if (!Properties.getParatroopersCanAttackDeepIntoEnemyTerritory(data.getProperties())) {
-        for (final Territory current :
-            CollectionUtils.getMatches(route.getMiddleSteps(), Matches.territoryIsLand())) {
-          if (Matches.isTerritoryEnemy(player).test(current)) {
-            return result.setErrorReturnResult("Must stop paratroops in first enemy territory");
-          }
-        }
+      if (!Properties.getParatroopersCanAttackDeepIntoEnemyTerritory(data.getProperties())
+          && route.getMiddleSteps().stream()
+              .anyMatch(Matches.territoryIsLand().and(Matches.isTerritoryEnemy(player)))) {
+        return result.setErrorReturnResult("Must stop paratroops in first enemy territory");
       }
     }
     return result;
   }
 
+  private Map<Unit, Unit> dependentsToUnitToTransportsMap(
+      final Map<Unit, Collection<Unit>> airTransportDependents, final MoveValidationResult result) {
+    Map<Unit, Unit> unitsToTransport = new HashMap<>();
+    for (Unit transport : airTransportDependents.keySet()) {
+      int capacity = TransportTracker.getAvailableCapacity(transport);
+      for (Unit beingTransported : airTransportDependents.get(transport)) {
+        int cost = beingTransported.getUnitAttachment().getTransportCost();
+        unitsToTransport.put(beingTransported, transport);
+        // Validate capacity, as airTransportDependents is coming from the move we're validating.
+        if (capacity < cost) {
+          result.setError("Not all units could be air transported");
```

Comment:
I wonder if we can make this error message more specific and helpful to the user. For example, could we say something like "not all units could be air transported, X out of Y transport capacity required"

Or, if we could say which units could not be transported, eg: "Not enough transport capacity for: X, Y, Z"

Reply:
Let me try using result.addDisallowedUnit() here instead, like it's done for the blitz case.

Reply:
Actually, addDisallowedUnit() is not an error, just a warning, so it's not the same semantics.

Thinking about this more, I think this is actually OK to leave as-is, because this situation won't happen in practice due to validation done in MovePanel - it doesn't let you load more units than allowed on an air transport. So this is just guarding against coding mistakes (e.g. if move panel changes or if AI submits invalid moves) or malicious players that are sending invalid moves with a custom client. As such, I'll just keep it as-is but add a comment to this extent.